Lock the package system while we're removing packages

isantop opened the pull request:
This does two things for us. 

First, obtaining a lock ensures that no other processes are using the package cache, and allows us to wait until any such processes have finished. 

Second, it prevents other processes from attempting to modify the package cache until after we have finished with removing the packages. 

Fixes #22

Commit:
fix(service): wait for package manager lock before trying to remove

Fixes #22

Commit:
fix: add more granular control over the apt package system

Instead of blindly hoping the package system is unlocked (and only making atomic changes between each call), we should instead obtain a process-wide lock over dpkg, perform our changes, and then only release the lock once we're finished with the package system. This allows us to wait for a package manager lock from another process to be released, obtain the lock, operate on the system with it locked (preventing changes by other processes), and then release the lock so that other processes can operate on the cache. This also allows us to display more granular status to the user, which can help them understand what the system is waiting on (and allowing them to take action if needed to speed the process along).

Fixes #22

Commit:
fix(apt): make module more fault-tolerant

We need to ensure that the package manager lock is released even in the event of a failure. Otherwise the package system may end up un-usable.

Commit:
refactor(service): remove now-unused dbus method

BREAKING CHANGE: We are no longer using this method, so no need to keep it around.

This code isn't being used anywhere else to my knowledge, but we should still increment the version as if it were just in case.

Commit:
release 1.0.1 → 1.1.0

isantop commented:
For testing this change, run `sudo apt upgrade` or `sudo apt install [package]` and leave the process running at the confirmation prompt. Then, open up Transition and ensure that after prompting for a password, that the status in the window states "Waiting for package manager lock". Then, complete the prompt and ensure that Transition begins removing the packages within 5 seconds. Finally, while Transition is removing packages, ensure that attempting to `sudo apt upgrade` either fails due to inability to obtain a lock, or that it retries until it can obtain a lock (after Transition is finished).

Commit:
fix(service): check for local lock before operating on the cache

Commit:
release 1.1.0 → 1.1.1

bflanagin reviewed (approved):
Works like a champ.

jacobgkau commented:
While testing this, I saw `Waiting for the package system lock` briefly, but then it proceeded to the next screen without removing, which isn't much different than the current behavior.

![Screenshot from 2021-02-11 14-00-08](https://user-images.githubusercontent.com/7199422/107698323-9dd4b800-6c71-11eb-9ec0-2835dcd966eb.png)

This was on 20.10 testing with just `chromium`. @bflanagin Did you see this?

jacobgkau commented:
Looks like waiting for the lock works when actually performing a package operation (such as installing a package), but not when updating the package lists. Is this intended? I think updating the package lists is what's more likely to happen in the background.

See my test case from #22:
> ...clear out `/var/lib/apt/lists/*`, then run an `apt update` and immediately try to remove the debs while the `apt update`'s still running.

isantop commented:
It looks like updating the apt cache doesn't put a lock on the package system. I can update the cache while another dpkg operation is in progress.

Commit:
fix(apt): obtain APT lock as well as dpkg lock

We need to obtain both an APT lock as well as a dpkg lock. This ensures that this is in place before starting any operations.

Fixes #22

Commit:
release 1.1.1 → 1.1.2

isantop commented:
@jacobgkau It looks like there needs to be a lock on both APT as well as the dpkg frontend. That should now be the case as of 50daccc

jacobgkau reviewed (approved):
Waiting for the lock now works when running `apt update`.

![Screenshot from 2021-02-12 13-51-47](https://user-images.githubusercontent.com/7199422/107821574-f5d6f180-6d39-11eb-91b9-f60c35345754.png)

Still works for `apt upgrade` too, and general regression testing passed on 20.04 and 20.10.
